File: backend/app/api/assets.py
"""Asset management endpoints."""

from fastapi import APIRouter, HTTPException, status
from sqlalchemy import select

from backend.app.schemas.asset import AssetCreate, AssetUpdate, AssetResponse
from backend.app.core.dependencies import DatabaseSession, CurrentUser
from backend.app.services.asset_service import create_asset, update_asset
from backend.app.models.asset import Asset


router = APIRouter(prefix="/assets", tags=["Assets"])


@router.post("", response_model=AssetResponse)
async def create_new_asset(
    asset_in: AssetCreate,
    db: DatabaseSession,
    current_user: CurrentUser
):
    """Create a new asset."""
    asset = await create_asset(db, asset_in, current_user)
    return asset


@router.get("", response_model=list[AssetResponse])
async def list_assets(
    db: DatabaseSession,
    current_user: CurrentUser,
    asset_type: str | None = None,
    skip: int = 0,
    limit: int = 100
):
    """List assets."""
    query = select(Asset)

    if asset_type:
        query = query.where(Asset.asset_type == asset_type)

    result = await db.execute(query.offset(skip).limit(limit))
    assets = result.scalars().all()
    return assets


@router.get("/{asset_id}", response_model=AssetResponse)
async def get_asset(
    asset_id: int,
    db: DatabaseSession,
    current_user: CurrentUser
):
    """Get asset by ID."""
    result = await db.execute(select(Asset).where(Asset.id == asset_id))
    asset = result.scalar_one_or_none()

    if not asset:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asset not found"
        )

    return asset


@router.patch("/{asset_id}", response_model=AssetResponse)
async def update_asset_endpoint(
    asset_id: int,
    asset_in: AssetUpdate,
    db: DatabaseSession,
    current_user: CurrentUser
):
    """Update an asset."""
    asset = await update_asset(db, asset_id, asset_in, current_user)
    return asset

File: frontend/src/api/assets.ts
import request from "@/utils/request";
import type { Asset } from "@/types/asset";

// 查询参数类型
interface GetAssetsParams {
  skip?: number;        // 跳过多少条（用于分页）
  limit?: number;       // 最多返回多少条（用于分页）
  asset_type?: string;  // 按资产类型筛选（可选）
}

// 获取资产列表，支持分页和类型筛选
export const getAssets = (params?: GetAssetsParams) =>
  request.get<Asset[]>("/assets", { params });

// 根据 ID 获取单个资产详情
export const getAsset = (id: number) =>
  request.get<Asset>(`/assets/${id}`);

// 创建新资产
// Partial<Asset> 表示 Asset 的所有字段都是可选的（不必填满）
export const createAsset = (data: Partial<Asset>) =>
  request.post<Asset>("/assets", data);

// 更新指定 ID 的资产（局部更新，只传要改的字段）
// PATCH 与 PUT 的区别：PATCH 只更新传入的字段，PUT 会替换整个对象
export const updateAsset = (id: number, data: Partial<Asset>) =>
  request.patch<Asset>(`/assets/${id}`, data);